### PR TITLE
WIP: Approach for API Versioning via config, headers

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR is a prototype of adding Content Header negotiation to our Pyramid app to support API versioning. It allows views to declare what versions they are available for.  As such, it:

* **Defines a list of known/published API versions as well as the current API version.**

    Right now, these are defined as non-override-able settings (`config.registry.settings[“api.versions”]`, `config.registry.settings[“api.version.current”]`. This is not a pattern we’ve followed before—having a setting that cannot be altered by env—so I’m more than open to putting this somewhere else. Basically these are constants that I’d like available during the app’s lifecycle.

   My goals are to make this easy to find for a future dev, and clear, not buried somewhere in a submodule that may make it confusing. That said, as long as this information—a list of all versions the app is aware of plus the current version—is available during bootstrap (i.e. `config`) and view execution (i.e.  `request)`, I’m cool.

* **Extends API view configuration such that views may declare which API versions they support**

    The `@api_config` decorator accepts a `versions`  (List) param. This param, for the moment, is optional and defaults to the current version of the API, but I’m leaning toward making it required so that we don’t end up with unexpected configurations. I waffled over this for a while this morning. Initially, I thought I only wanted us to have to set `versions` config on views that were not for the default version—to make this config more implicit than explicit. However, as I mused over how longer-term maintenance would go, it seems like just taking a few minutes to add this param to all the existing API views and be done with it and have really clear config going on. Anyway.

* **Assigns the right `accepts` config to a view during config, based on its versions**

   If a view supports the current version of the API, it will be configured with `accept=“application/json` as well as the Accept header for the current API version (e.g. `accept=“application/vnd.hypothesis.v1+json”`). If a declared version is unknown, the config will raise at present.
* **Weights `Accept` headers such that views for the current version of the API will be preferred if a request’s Accept header is missing, invalid or generic**

    This allows us to use the current API version’s view when a request lacks a versioned Accept header (for example, a missing Accept or an `Accept:application/json`)

    Note: If a request has an *invalid* Accept header, it’ll raise a 404. This surprised me because [Pyramid's documentation](https://docs.pylonsproject.org/projects/pyramid/en/latest/narr/viewconfig.html#accept-content-negotiation) on this makes it sound like the heaviest-weighted accepts-view will get used if the header is invalid. In any case, this is not a regression: if you send an invalid Accept at our API endpoints today, you’ll get a 404 (a missing Accept is OK and works, but not an invalid one). Shrug?
* **Does a light/naive refactoring of the way that API services are registered as “links”** (i.e. what gets returned by the API’s index route). 

    This was mostly done to “make things work” for the purposes of a speedy prototype.
* **Adds a couple example “v2” views for the purpose of illustration**

    If you’re interested, you can run this branch and try hitting either `GET /api/groups` or `GET /api` locally with different Accept headers and see what response you get. I guess leaving this as an exercise for the reader is rude, so I’ll spill the beans:

    * Send `Accept:application/json`, `Accept:application/vnd.hypothesis.v1+json` or no Accept header and you’ll get the default/v1 response
    * Send `Accept:application/vnd.hypothesis.v2+json` and you’ll get the “new”, fake v2 response on those endpoints

### So…

I’m looking for feedback on the general approach, things like where to stick settings and whether the config stuff seems generally OK. Don’t get too caught up on the lack of tight factoring on the config functions themselves as I will be refactoring when I do the actual implementation. I just wanted to get a prototype out quickly. More notes in code comments to come…

### CODE REVIEW

File: h/config.py [hunk 1]
```diff
@@ -143,6 +143,11 @@ def configure(environ=None, settings=None):
     es_logger = logging.getLogger("elasticsearch")
     es_logger.addFilter(ExceptionFilter((("ReadTimeoutError", "WARNING"),)))
 
+    # Add API version settings
+    # These cannot be overridden by env
+    settings["api.versions"] = ["v1", "v2"]
```
Comment (lyzadanger) [inline h/config.py]: Using simple strings (e.g. `"v2"`) makes these real easy to work with in various parts of the code. Can y'all think of any reason that not using semver formatting here could bite us?

File: h/views/api/__init__.py [hunk 2]
```diff
@@ -3,4 +3,13 @@
 
 
 def includeme(config):
+    config.add_accept_view_order("application/json")
```
Comment (lyzadanger) [inline h/views/api/__init__.py]: You'll note zero tests for this yet. Two reasons: speed of getting this PR out and uncertainty over how best to test, so I'll ask here! Is this something that needs to be tested with functests? Is there another way to ask Pyramid, in test, whether a given request matches a given view? I re-read Pyramid's testing docs and nothing jumped out.

Comment (lyzadanger) [reply h/views/api/__init__.py]: Oh, also, I don't love burying this config in `h.views.api`'s `__init__`, but it _does_ follow existing patterns. More than happy to move this code somewhere else.

File: h/views/api/config.py [hunk 3]
```diff
@@ -4,6 +4,12 @@
 from h.util import cors
 
 
+class APIConfigError(Exception):
```
Comment (lyzadanger) [inline h/views/api/config.py]: Tossed this custom Exception in here but have zero attachment to the methodology of raising in this module. Looking for opinions. /cc @seanh

File: h/views/api/config.py [hunk 4]
```diff
@@ -51,24 +63,45 @@ def add_api_view(
...
+    known_versions = config.registry.settings["api.versions"]
+    view_versions = versions or [current_version]
+
+    registry = config.registry
+    if not hasattr(registry, "api_links"):
+        registry.api_links = {}
+
+    for version in view_versions:
+        if version not in known_versions:
+            raise APIConfigError(
```
Comment (lyzadanger) [inline h/views/api/config.py]: I didn't bother formatting/defining this Exception carefully yet until I'm sure what kind of Exception-raising we wanna do here.

File: h/views/api/config.py [hunk 5]
```diff
@@ -51,24 +63,45 @@ def add_api_view(
...
-        registry = config.registry
-        if not hasattr(registry, "api_links"):
-            registry.api_links = []
-        registry.api_links.append(link)
+    current_version = config.registry.settings["api.version.current"]
+    known_versions = config.registry.settings["api.versions"]
+    view_versions = versions or [current_version]
+
+    registry = config.registry
+    if not hasattr(registry, "api_links"):
```
Comment (lyzadanger) [inline h/views/api/config.py]: I plan on refactoring this links stuff entirely during implementation.

File: h/views/api/config.py [hunk 6]
```diff
@@ -51,24 +63,45 @@ def add_api_view(
...
+                registry.api_links[version] = []
+            registry.api_links[version].append(link)
 
     config.add_view(view=view, **settings)
+
+    # Add version-specific Accept headers. This requires re-adding the view
+    # as `add_view` only takes one value for `accept` at a time
+    for version in view_versions:
+        settings["accept"] = "application/vnd.hypothesis." + version + "+json"
+        config.add_view(view=view, **settings)
```
Comment (lyzadanger) [inline h/views/api/config.py]: Hokay. This was the most annoying part to figure out. Yeah, you an call `config.add_view` multiple times on the same view, but I can't just do `config.add_view(view=view, accept="relevant_accept_header"); I have to pass on the `settings` from the first call to `add_view` or it I get conflict errors. It's possible that only a subset of those `settings` are needed to extend an existing view's config, but it wasn't obvious what and I didn't want to take the time to futz around with that. So: looking for any input on this. Maybe it's fine as-is? It's sort of a Pyramid oddity that one cannot extend existing config, only "re-add" the view in entirety. Shrug.

File: tests/h/views/api/index_test.py [hunk 7]
```diff
@@ -211,3 +206,11 @@ def test_it_returns_the_right_links_for_user_endpoints(
         assert links["user"]["update"]["url"] == (host + "/dummy/users/:username")
 
         assert set(links["user"].keys()) == set(["create", "update"])
+
+
+@pytest.fixture
```
Comment (lyzadanger) [inline tests/h/views/api/index_test.py]: This is an abomination of a hack. These view tests need to be completely refactored—most of them need to be moved into the tests for `h.views.api.config` instead of being tests for the view. The config stuff here should be mocked. Anyway, will fix during implementation.

File: h/views/api/config.py [hunk 8]
```diff
@@ -51,24 +63,45 @@ def add_api_view(
...
-            "description": description,
-        }
-
-        registry = config.registry
-        if not hasattr(registry, "api_links"):
-            registry.api_links = []
-        registry.api_links.append(link)
+    current_version = config.registry.settings["api.version.current"]
+    known_versions = config.registry.settings["api.versions"]
+    view_versions = versions or [current_version]
```
Comment (lyzadanger) [inline h/views/api/config.py]: In my current plan of approach, `versions` will not remain an optional param.

**Comment by lyzadanger** [comment]
Oh, dear, the failing functional tests on Jenkins are a head-scratcher. Those tests are passing just fine for me locally. The way they're failing makes me wonder if this is an environment/settings thing, possibly? I may need some help sleuthing that out. The failures suggest that, on Jenkins, current API version setting isn't getting set (and so no view gets assigned `application/json` and so view matching is random if there are two possible matches for the same route). The manifestation, in any case, is that in at least some cases on Jenkins during functional tests, v2 views are matching even though no Accept header is set.

**Comment by Ahmad9334** [comment]
h/config.py
@@ -143,6 +143,11 @@ def configure(environ=None, settings=None):
    es_logger = logging.getLogger("elasticsearch")
    es_logger.addFilter(ExceptionFilter((("ReadTimeoutError", "WARNING"),)))

     # Add API version settings
    # These cannot be overridden by env
    settings["api.versions"] = ["v1", "v2"]
@lyzadanger  lyzadanger • about 8 hours ago
Using simple strings (e.g. "v2") makes thes

File: h/config.py (hunk 1 above)
Comment (seanh) [reply h/config.py]: I think simple strings are probably best. If we've made incompatible API changes, equivalent to when you would increment the major version number in SemVer, then we'll increment the version number here. If we've only added new functionality (would increment the minor version number in SemVer) or made bug fixes (patch number in SemVer) we'll just deploy those and won't make any API version number change. I think this makes sense as I don't think we'd want to version the API (and keep on supporting those previous versions if request by an `Accept` header??) every time we deploy fix a bug or add a new endpoint or optional argument.

I think SemVer is mostly for dependencies (e.g. Python and JavaScript packages) and not for APIs (or for apps like h, bouncer, via, lms and client). (semver.org itself claims that the benefits are to do with making "dependency management" easier).

The one problem that comes to mind is if someone was writing an h client that was going to work with multiple different instances of h out there in the wild. If all those instances of h were on API v3, but they were all on different versions of v3, so some had certain new features and bug fixes and some didn't, it might be difficult for the client to know which features it could use with which instance of h. Maybe this would be discoverable by handling 4xx responses and the like.

Anyway, we're not in a world of multiple instances of h and having to update a patch or minor version number every time we make an API change sounds like a real pain

File: h/config.py [hunk 9]
```diff
@@ -143,6 +143,11 @@ def configure(environ=None, settings=None):
...
+    settings["api.version.current"] = "v1"
```
Comment (seanh) [inline h/config.py]: I wonder if `"api.version.default"` would be a better name? Since I think this is the default version that you get if you don't request a particular version in an `Accept` header.

Do I understand correctly that these two settings cannot be changed either by INI files or by envvars? So if we wanted to change either one of these settings, we'd have to change the Python code and redeploy? In that case I think these should actually be `CONSTANTS` not settings.

If the suggestion is amenable here's some ideas for where to put constants:

1. If they're only used within `h.views.api` make them private constants somewhere in that package:

   1. If only needed in one `h/views/api/*.py` module then make them `_CONSTANTS` in that module
   1. If they're needed throughout the `h.views.api` package, put them as private `_CONSTANTS` in `h/views/api/__init__.py` or, if that would cause a circular import, in `h/views/api/_constants.py` (and then `__init__.py` and any other modules that need them import them from `_constants.py`)

1. Or if they're potentially needed throughout the entire app then make them public `CONSTANTS` as part of the `h.views.api` public API. They would go in `h/views/__init__.py` unless they had to be moved into `h/views/_constants.py` to avoid circular imports. Either way `__init__.py` would add them to its `__all__` in order to make them part of its public API.

   Any code throughout the app could then just do `from h.views.api import API_VERSION_CURRENT`.

File: h/views/api/__init__.py (hunk 2 above)
Comment (seanh) [reply h/views/api/__init__.py]: Hmmmmm.. Doesn't this accept view ordering apply to the entire app, not just the API views? If so it might belong in `app.py`

Comment (seanh) [reply h/views/api/__init__.py]: There isn't a builtin way to write a "test that it called a given view" test in Pyramid. I believe that one could be devised. We could write a tween or other middleware or hook-type thingy that's used only in the tests and that plugs into Pyramid's view lookup machinery and just records which view was chosen. Tests could then use this tween to assert about which view was chosen.

I'm dubious about the value of it though and wonder whether a functional test would be better.

File: h/views/api/config.py (hunk 3 above)
Comment (seanh) [reply h/views/api/config.py]: I guess this would only ever be raised if the code was wrong (an `@api_config` had an invalid version number) and it would be raised at application startup? So code that raises this would never get committed and deployed (since the app would be failing to even start up).

In that case I think you could probably just raise `ValueError` instead of a custom exception.

If you still want a custom exception (for example because it makes writing tests easier) I think it should go in `h/views/api/exceptions.py` just for the sake of following the pattern of putting exceptions into `exceptions.py`. It'd make sense as a `ValueError` subclass to rather than just an `Exception` subclass, but it doesn't matter much.

File: h/views/api/config.py (hunk 6 above)
Comment (seanh) [reply h/views/api/config.py]: I'm not sure why Pyramid's `accept=` doesn't accept a list of strings :man_shrugging: But given that it doesn't I wonder whether the `api_version` setting should also be a single string not a list of strings?

File: tests/h/views/api/index_test.py (hunk 7 above)
Comment (seanh) [reply tests/h/views/api/index_test.py]: I wonder of the existing `pyramid_config` fixture could be used?

File: h/views/api/config.py (hunk 6 above)
Comment (lyzadanger) [reply h/views/api/config.py]: I want views to be able to support multiple versions (e.g. `versions=["v1", "v2"]`)...

Comment (lyzadanger) [reply h/views/api/config.py]: (I feel like I might be mis-reading your comment...can you rephrase, actually?)

**Comment by seanh** [comment]
I think making `versions` explicitly required in every API view config, rather than having it implicitly default to the current version, sounds like a good idea

File: h/views/api/__init__.py (hunk 2 above)
Comment (lyzadanger) [reply h/views/api/__init__.py]: > Doesn't this accept view ordering apply to the entire app, not just the API views? If so it might belong in `app.py`

Generally, I'd say...sure, okay. But that would mean that `app.py` would need to be privy to the versioning constants we're talking about putting in `h.views.api` somewhere...

File: h/views/api/config.py (hunk 6 above)
Comment (seanh) [reply h/views/api/config.py]: As in:

```python
@api_config(... version="v1")
@api_config(... version="v2")
def my_view(...):
    ...
```

**Comment by lyzadanger** [comment]
Thanks for the great feedback! Closing now and starting implementation phase in small chunks...
